### ISSUE TITLE
Default chunks for the zarr store

### ISSUE THREAD

**Issue by SutarPrasad** [openghg/openghg#908]
### What is your issue?

Need to have the default chunk value set for the Zarr storage move. 
Especially for `co2` data files it is required to at least chunk the time dimension.

As of the current investigation having the default chunk value set to `chunks = {time:100}` is the optimum value for the `standardise_*` functions.

**Comment by SutarPrasad** [comment]
Linked to https://github.com/openghg/openghg/pull/803

**Comment by brendan-m-murphy** [comment]
This is the code from the chunk tests I ran during the storage sprint:

```
def load_fp_mf_dataset(
    file_paths: list[str],
    chunks: Optional[dict[str, int]] = None,
    time_chunk: Optional[int] = None,
    lat_chunk: Optional[int] = None,
    lon_chunk: Optional[int] = None,
    max_chunk_size: Optional[Union[int, float]] = None,
    max_chunk_by: list[str] = ["lat", "lon"],
    **open_mfds_kwargs,
) -> xr.Dataset:
    """Load footprints from list of file paths using xr.open_mfdataset.

    Args:
        file_paths: list of file paths as strings
        chunks: optional dictionary of chunk sizes; keys must be dims of the netCDF to be loaded
        time_chunk: optional chunk size for `time` dim
        lat_chunk: optional chunk size for `lat` dim
        lon_chunk: optional chunk size for `lon` dim
        max_chunk_size: max number of bytes in chunk of `fp` variable (assuming no compression)
        max_chunk_by: list of dimensions to scale to reduce chunksize, default = ["lat", "lon"],
        open_mfds_kwargs: keyword args to pass to xr.open_mfdataset, e.g. {'parallel': True}

    Returns:
        xr.Dataset wrapping dask arrays for given files, with specified chunks
    """
    with xr.open_dataset(file_paths[0]) as ds:
        dim_sizes = dict(ds.sizes)
        fp_dtype = str(ds.fp.dtype)

    if any(dim not in dim_sizes for dim in ["time", "lat", "lon", "height"]):
        raise ValueError(
            f'File {file_paths[0]} is missing one of the following dimensions: "time", "lat", "lon", "height"'
        )

    # make 'chunks' dict, using dim_sizes for any unspecified dims
    if chunks is None:
        chunks = dim_sizes.copy()
    else:
        chunks = dict(dim_sizes, **chunks)

    # any individually specified chunks take precedence
    if time_chunk:
        chunks["time"] = time_chunk
    if lat_chunk:
        chunks["lat"] = lat_chunk
    if lon_chunk:
        chunks["lon"] = lon_chunk

    # chunk along lat/lon to get below min size
    if max_chunk_size:
        if m := re.search(r"\d+$", fp_dtype):
            fp_dtype_bytes = int(m.group(0)) / 8
        else:
            fp_dtype_bytes = 8  # assume worst case: 64 bit

        current_chunk_size = fp_dtype_bytes * chunks["time"] * chunks["lat"] * chunks["lon"]

        if current_chunk_size > max_chunk_size:
            ratio = np.power(max_chunk_size / current_chunk_size, 1 / len(max_chunk_by))
            for dim in max_chunk_by:
                # rescale chunks, but don't allow chunks smaller than 10
                chunks[dim] = max(int(ratio * chunks[dim]), 10)

    # chunks smaller than dims should be set after loading
    small_chunks = {}
    for k in dim_sizes:
        if chunks[k] < dim_sizes[k]:
            small_chunks[k] = chunks.pop(k)

    if small_chunks:
        result = xr.open_mfdataset(file_paths, chunks=chunks, **open_mfds_kwargs).chunk(small_chunks)
    else:
        result = xr.open_mfdataset(file_paths, chunks=chunks, **open_mfds_kwargs)

    return result
```

This is specialised to footprints, but we could make it more general by computing the max chunk size by iterating over data variables and multiplying the size of the dimensions of the coordinates for each data variable.

The key equation is: 
```
current_chunk_size = fp_dtype_bytes * chunks["time"] * chunks["lat"] * chunks["lon"]
```
i.e. the size of a chunk in memory is the "size" of the array (product of the lengths of the coordinates) times the size of the data type (in bytes).

**Comment by brendan-m-murphy** [comment]
The Dask docs say that the number of chunks in memory will often be *twice* the number of threads: https://docs.dask.org/en/latest/array-best-practices.html?highlight=OMP_NUM_THREADS

So if we're running batch scripts on Blue Pebble, we need `(RAM) > 2 x (number of threads) x (max chunk size)`, and for a Macbook pro, if we don't restrict the number of threads, then we'll have 16 available, so if we have 16GB ram, we need to have chunk sizes below 500MB to be safe.

The Dask docs suggest keeping chunks above 100MB. So I think we're aiming for 100 - 500 MB per chunk (in memory; the chunks will be compressed on disk).

For a high time res. footprint, there are 4 dimensions: time, lat, lon, and H_back, and the data type seems to be float32. The size of H_back is 25, so 

```
(size H_back) x (size of data type in bytes) = 100 bytes
```

Thus we need 
```
1e6 < (time) x (lat) x (lon) < 5e6
```
(ignoring the difference between 10^6 and 2^20...)

The size of `(lat) x (lon)` is  293 * 391 = 114,563 or about 1.1e5, so we need `9 < (time) < 45` if we don't chunk lat and lon.

**Comment by brendan-m-murphy** [comment]
Alternatively, we could have `H_back` chunk sizes of 1 and use the same chunk sizes for time, lat, and lon as for regular footprints.

**Comment by rt17603** [comment]
> Alternatively, we could have `H_back` chunk sizes of 1 and use the same chunk sizes for time, lat, and lon as for regular footprints.

When calculating the modelled observations, each footprint in the `H_back` dimension will be combined with the appropriate emissions before being summed. As an operation that sounds like that would actually work ok if this was split along the `H_back` dimension.

Perhaps would be good footprint_data_merge for co2 with this as as we know combining co2 footprints is slow anyway and wouldn't want to make this slower.

**Comment by SutarPrasad** [comment]
While looping through the BSD_2015_co2 files following error was generated:

**code**:
for file in new_co2_files: 
    standardise_footprint(filepath=file,
                      site="BSD",
                      model="NAME",
                      domain="EUROPE",
                      inlet="250magl",
                      chunks={"time":100, "lat":100, "lon":100, "height":100},
                      )

**Error**:
`"ValueError: conflicting sizes for dimension 'time': length 744 on 'fp_HiTRes' and length 1464 on {'H_back': 'H_back', 'time': 'PBLH', 'lat': 'fp', 'lon': 'fp'}`

**Comment by rt17603** [comment]
Closed as part pf #831
